### ISSUE TITLE
Optional dependencies

### ISSUE THREAD

**Issue by rmalouf** [nltk/nltk#416]
According to the docs [here](http://nltk.org/install.html), nltk requires pyyaml and has an optional dependency on numpy.  Looking through the code, though, there are a few other packages that get imported (matplotlib, scipy, sklearn, svmlight, unidecode) that should maybe be mentioned as optional dependencies.  But more importantly, when one of these optional packages are missing, attempts to use parts of nltk will sometimes fail with a possibly confusing error message, e.g.:

```
Traceback (most recent call last):
  File "nltk/classify/maxent.py", line 1484, in <module>
    demo()
  File "nltk/classify/maxent.py", line 1481, in demo
    classifier = names_demo(MaxentClassifier.train)
  File "/home1/malouf/nltk/nltk/classify/util.py", line 184, in names_demo
    classifier = trainer( [(features(n), g) for (n,g) in train] )
  File "nltk/classify/maxent.py", line 288, in train
    train_toks, trace, encoding, labels, **cutoffs)
  File "nltk/classify/maxent.py", line 1111, in train_maxent_classifier_with_iis
    empirical_ffreq = (calculate_empirical_fcount(train_toks, encoding) /
  File "nltk/classify/maxent.py", line 1067, in calculate_empirical_fcount
    fcount = numpy.zeros(encoding.length(), 'd')
NameError: global name 'numpy' is not defined
```

Missing dependencies should probably be handled more clearly and consistently.  But, some questions:
- Should some of these dependencies be made non-optional? I'm thinking mostly of matplotlib and numpy.  Many python distributions come with them pre-installed.
- Should the remaining optional dependencies be added to `setup.py` as "extras", as described [here](http://pythonhosted.org/distribute/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies), to encourage users to install them?
- For missing dependencies, is it better to fail at import time or at run time? 
- How fancy do we want to get with the error message?  Simply state that a package is missing? Point the user to a webpage describing how to install NLTK?

For reference, here are the source files that import specific packages:
#### pylab

./nltk/probability.py
./nltk/parse/dependencygraph.py
./nltk/draw/dispersion.py
./nltk/app/wordfreq_app.py
./nltk/app/**init**.py
./nltk/tokenize/texttiling.py
#### numpy

./nltk/classify/maxent.py
./nltk/classify/megam.py
./nltk/classify/scikitlearn.py
./nltk/classify/tadm.py
./nltk/metrics/segmentation.py
./nltk/tag/hmm.py
./nltk/test/classify_fixt.py
./nltk/test/classify_fixt.py
./nltk/test/probability_fixt.py
./nltk/test/unit/test_tag.py
./nltk/test/unit/test_hmm.py
#### scipy

./nltk/classify/scikitlearn.py
./nltk/metrics/scores.py
#### sklearn

./nltk/classify/scikitlearn.py
#### unidecode

./nltk/compat.py
#### svmlight

nltk/classify/svm.py

**Comment by kmike** [comment]
Hi @rmalouf,

A good ticket and good questions. My opinion:

> Should some of these dependencies be made non-optional? I'm thinking mostly of matplotlib and numpy. Many python distributions come with them pre-installed.

-1. I think no, they shouldn't be made non-optional. Matplotlib is used for drawing - I think it is mostly useful for interactive work or for learning; it may not be needed in many production environments. Numpy is still not easy to install to virtualenv, and a lot of NLTK modules doesn't need it. 

Without these dependencies NLTK can be installed and used as a pure-python module, without a C/C++/Fortran compiler, which is very good in my opinion.

By the way, I was about to propose making PyYAML optional. And maybe removing / moving to nltk_contrib some of the classifiers - it seems that nobody knows how to install megam + svmlight bindings has bugs, licensing restrictions (non-commercial usage) and great alternatives in scikit-learn.

> Should the remaining optional dependencies be added to setup.py as "extras", as described here, to encourage users to install them?

-0. I'm slightly against this because it drives NLTK's setup.py further from standard library: extras_require is a non-standard option (distutils from standard library doesn't have it). NLTK uses distribute so this is really OK. By the way, distribute will be deprecated soon (it was merged with setuptools if I'm not mistaken).

> For missing dependencies, is it better to fail at import time or at run time?

I'd say at run time, because 

1) the dependencies are often required only for a small subset of module functionality (and even if a subset is large some functions or classes from a module can still be used if we fail at run time);
2) NLTK often imports submodules in package's `__init__.py`; missing dependency shouldn't cause an import error for a whole package; not importing a module in case of missing dependency is bad because error messages are not informative in such cases. 

> How fancy do we want to get with the error message? Simply state that a package is missing? Point the user to a webpage describing how to install NLTK?

+1 for "simply state that a package is missing" (e.g. by raising an appropriate ImportError). There are downsides in being too smart: webpage referred from older NLTK version could disappear; packaging is changed and install instructions contained in an error message could also become outdated. I think ImportErrors are quite standard and even novice user could figure out that if code fails with "numpy" ImportError then numpy package should be installed. Then they google "install numpy" and get an up-to-date instruction on how to install numpy.

> unidecode
> ./nltk/compat.py

There is also text-unidecode soft dependency (text-unidecode is a better licensed but poorer implemented unidecode alternative).

**Comment by alvations** [comment]
3+ years on, the new `pip` installation should handle this issue c.f. #1538

**Comment by alvations** [comment]
Resolved in  #1538
Closing resolved issues to prepare for v3.2.5
